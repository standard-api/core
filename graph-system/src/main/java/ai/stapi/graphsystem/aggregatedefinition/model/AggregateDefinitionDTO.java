package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.ArrayList;
import java.util.List;

public class AggregateDefinitionDTO {

  private String id;
  private String name;
  private String description;
  private List<CommandHandlerDefinitionDTO> command;
  private StructureDefinitionId structure;

  protected AggregateDefinitionDTO() {
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      List<CommandHandlerDefinitionDTO> command,
      StructureDefinitionId structure
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.command = command;
    this.structure = structure;
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      StructureDefinitionId structure
  ) {
    this(id, name, description, new ArrayList<>(), structure);
  }

  public AggregateDefinitionDTO addCommandHandlerDefinition(
      CommandHandlerDefinitionDTO commandHandlerDefinition
  ) {
    this.command.add(commandHandlerDefinition);
    return this;
  }

  @JsonIgnore
  public boolean containsOperation(String operationId) {
    return this.command.stream().anyMatch(
        someCommand -> someCommand.getOperation().getId().equals(operationId)
    );
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  public List<CommandHandlerDefinitionDTO> getCommand() {
    return command;
  }

  public StructureDefinitionId getStructure() {
    return structure;
  }
}