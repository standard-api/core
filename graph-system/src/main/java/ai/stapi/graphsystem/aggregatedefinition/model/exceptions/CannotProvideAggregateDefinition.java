package ai.stapi.graphsystem.aggregatedefinition.model.exceptions;

public class CannotProvideAggregateDefinition extends RuntimeException {

  public CannotProvideAggregateDefinition(String message) {
    super(message);
  }

  private CannotProvideAggregateDefinition(
      String aggregateName,
      String becauseMessage,
      Throwable cause
  ) {
    super(
        String.format(
            "Cannot Provide Aggregate Definition for: '%s', because %s%nCause: %s",
            aggregateName,
            becauseMessage,
            cause.getMessage()
        )
    );
  }

  public static CannotProvideAggregateDefinition becauseItDoesNotExist(String aggregateName) {
    return new CannotProvideAggregateDefinition(
        String.format(
            "Cannot Provide Aggregate Definition for: '%s', because it does not exists.",
            aggregateName
        )
    );
  }

  public static CannotProvideAggregateDefinition becauseRelatedOperationDefinitionNotFound(
      String aggregateName,
      String operationName,
      Throwable cause
  ) {
    return new CannotProvideAggregateDefinition(
        aggregateName,
        String.format("related operation with name '%s' was not found.", operationName),
        cause
    );
  }

  public static CannotProvideAggregateDefinition becauseRelatedStructureSchemaNotFound(
      String aggregateName,
      Throwable cause
  ) {
    return new CannotProvideAggregateDefinition(
        aggregateName,
        "related Structure Schema was not found.",
        cause
    );
  }


  public static CannotProvideAggregateDefinition becauseThereIsNoCommandHandlerForOperation(
      String operationDefinitionId
  ) {
    return new CannotProvideAggregateDefinition(
        String.format(
            "Cannot Provide Aggregate Definition for: '%s', because there is no command handler for this operation.",
            operationDefinitionId
        )
    );
  }
}
