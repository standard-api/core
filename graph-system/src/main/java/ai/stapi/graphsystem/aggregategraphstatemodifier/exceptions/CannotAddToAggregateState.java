package ai.stapi.graphsystem.aggregategraphstatemodifier.exceptions;

import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.schema.structureSchema.ComplexStructureType;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class CannotAddToAggregateState extends RuntimeException {

  private CannotAddToAggregateState(String becauseMessage) {
    super("Cannot add to aggregate state, because " + becauseMessage);
  }

  public CannotAddToAggregateState(List<CannotAddToAggregateState> cannotAddToAggregateStateList) {
    super(
        cannotAddToAggregateStateList.stream()
            .map(Throwable::getMessage)
            .collect(Collectors.joining("\n"))
    );
  }

  public static CannotAddToAggregateState becauseModificationPathIsInvalid(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification path present in modification definition was invalid.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseModificationPathDoesNotStartWithCorrectAggregateType(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition,
      String actualAggregateType
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification path does not start with correct aggregate type.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'%n" +
                "Actual aggregate type: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath(),
            actualAggregateType
        )
    );
  }

  public static CannotAddToAggregateState becauseModificationPathContainsFieldNameNotPresentInSchema(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification path present in modification definition "
                + "contains field name not present in schema.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseModificationPathContainsFieldNameWhichIsPrimitiveButItShouldNot(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification path present in modification definition "
                + "contains field name which is of primitive type, but it is not end of path.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseModificationDefinitionDoesNotHaveInputValueParameterNameSpecified(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification definition does not have input value parameter name specified.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseModifiedNodeSpecifiedByModificationPathWasNotPresent(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition
  ) {
    var modificationPath = modificationDefinition.getModificationPath();
    var split = modificationPath.split("\\.");
    return new CannotAddToAggregateState(
        String.format(
            "node to be modified (specified by modification path) was not present in aggregate state.%n"
                +
                "Operation name: '%s'%nModification kind: '%s'%nExpected present parent node path: %s%nWhole Modification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            String.join(".", Arrays.copyOfRange(split, 0, split.length - 1)),
            modificationPath
        )
    );
  }

  public static CannotAddToAggregateState becauseThereAreIndistinguishableNodes(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition,
      String expectedIdSourcePath
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification definition has path which leads through list. %s%n%s%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            "But there is no id to distinguish between the items.",
            String.format(
                "Maybe invalid command handler definition? You can fix this by adding parameter of type id to operation definition with source path: '%s'.",
                expectedIdSourcePath
            ),
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseThereAreMoreWaysToDistinguisNodes(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationStructureType,
      String sourcePathOfId
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "modification definition has path which leads through multiple nodes. %s%n" +
                "Operation name: '%s'%nModification kind: '%s'%nModification path: '%s'",
            String.format(
                "But there are multiple parameters on operation definition with source path: '%s'.",
                sourcePathOfId),
            operationStructureType.getDefinitionType(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseThereIsNoNodeWithIdSpecifiedAtSourcePath(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationStructureType,
      String sourcePathOfId,
      UniqueIdentifier id
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "there is no node with id specified at source path.%n" +
                "Operation name: '%s'%nNode path: '%s'%nId: '%s'%nModification kind: '%s'%n"
                + "Modification path: '%s'",
            operationStructureType.getDefinitionType(),
            sourcePathOfId,
            id.getId(),
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseThereIsNoNodeWithIdSpecifiedAtStartIdParameterName(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationStructureType,
      String sourcePathOfId,
      UniqueIdentifier id,
      String startIdParameterName
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "there is no node with id specified at start id parameter name.%n" +
                "Operation name: '%s'%nNode path: '%s'%nId: '%s'%nModification kind: '%s'%n"
                + "Start id parameter name: '%s'",
            operationStructureType.getDefinitionType(),
            sourcePathOfId,
            id.getId(),
            modificationDefinition.getKind(),
            startIdParameterName
        )
    );
  }

  public static CannotAddToAggregateState becauseThereIsNoSourcePathAtStartIdParameterName(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationStructureType,
      UniqueIdentifier id,
      String startIdParameterName
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "there is no node with source path specified at start id parameter name.%n" +
                "Operation name: '%s'%nId: '%s'%nModification kind: '%s'%n"
                + "Start id parameter name: '%s'",
            operationStructureType.getDefinitionType(),
            id.getId(),
            modificationDefinition.getKind(),
            startIdParameterName
        )
    );
  }

  public static CannotAddToAggregateState becauseThereIsNoIdInCommandAtStartIdParameterName() {
    return new CannotAddToAggregateState(
        "there is no id in command at start id parameter name. " +
            "Should not ever happen, bcs validator already should prevent this."
    );
  }

  public static CannotAddToAggregateState becauseThereAreEdgesOnPathEvenThoughtThereShouldBeMaxOne(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition,
      String multipleEdgesPath
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "there are multiple edge on path event thought ther should be max one by schema.%n" +
                "Operation name: '%s'%nModification kind: '%s'%nMultiple edges path: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            modificationDefinition.getKind(),
            multipleEdgesPath,
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseThereAreIsNodeEdgeOnPathEvenThoughtThereShouldBeOne(
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationDefinition,
      String multipleEdgesPath
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "the node you are trying to modify should already exist. Please add it first with appropriate command.%n"
                +
                "Operation name: '%s'%nMissing node path: '%s'%nModification kind: '%s'%nModification path: '%s'",
            operationDefinition.getDefinitionType(),
            multipleEdgesPath,
            modificationDefinition.getKind(),
            modificationDefinition.getModificationPath()
        )
    );
  }

  public static CannotAddToAggregateState becauseStartIdIsNotOfCorrectFormat(
      String startIdValue,
      String startIdParameterName,
      EventFactoryModification modificationDefinition,
      ComplexStructureType operationStructureType
  ) {
    return new CannotAddToAggregateState(
        String.format(
            "start id found at start id parameter name is not of correct format (Id/Type).%n" +
                "Operation name: '%s'%%nModification kind: '%s'%n"
                + "Start id parameter name: '%s' Start id value: %s",
            operationStructureType.getDefinitionType(),
            modificationDefinition.getKind(),
            startIdParameterName,
            startIdValue
        )
    );
  }
}
