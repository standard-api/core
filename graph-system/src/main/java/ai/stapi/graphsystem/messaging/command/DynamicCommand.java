package ai.stapi.graphsystem.messaging.command;

import ai.stapi.identity.UniqueIdentifier;
import java.util.Map;

public class DynamicCommand extends AbstractPayloadCommand<UniqueIdentifier> {

  public static final String SERIALIZATION_TYPE_FIELD_NAME = "serializationType";
  public static final String TARGET_IDENTIFIER_FIELD_NAME = "targetIdentifier";

  private DynamicCommand() {
  }

  public DynamicCommand(
      UniqueIdentifier targetIdentifier,
      String serializationType,
      Map<String, Object> data
  ) {
    super(serializationType, targetIdentifier, data);
  }
}
