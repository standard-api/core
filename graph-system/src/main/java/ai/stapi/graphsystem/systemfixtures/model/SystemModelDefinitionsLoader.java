package ai.stapi.graphsystem.systemfixtures.model;

import ai.stapi.schema.adHocLoaders.FileLoader;
import ai.stapi.schema.adHocLoaders.AbstractFileModelDefinitionsLoader;
import ai.stapi.schema.scopeProvider.ScopeOptions;
import org.springframework.stereotype.Service;

@Service
public class SystemModelDefinitionsLoader extends AbstractFileModelDefinitionsLoader {

  public static final String SCOPE = "system";
  public static final String TAG = ScopeOptions.DOMAIN_TAG;

  public SystemModelDefinitionsLoader(
      FileLoader fileLoader
  ) {
    super(fileLoader, SCOPE, TAG);
  }
}
