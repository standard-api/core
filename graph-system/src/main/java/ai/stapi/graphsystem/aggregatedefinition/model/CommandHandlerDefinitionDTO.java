package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.graphsystem.eventdefinition.EventMessageDefinitionData;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;
import java.util.List;
import java.util.UUID;
import org.jetbrains.annotations.Nullable;

public class CommandHandlerDefinitionDTO {

  private OperationDefinitionDTO operation;
  private String creationalPolicy;
  private List<EventFactory> eventFactory;

  private CommandHandlerDefinitionDTO() {
  }

  public CommandHandlerDefinitionDTO(
      OperationDefinitionDTO operation,
      String creationalPolicy,
      List<EventFactory> eventFactory
  ) {
    this.operation = operation;
    this.creationalPolicy = creationalPolicy;
    this.eventFactory = eventFactory;
  }

  public OperationDefinitionDTO getOperation() {
    return operation;
  }

  public String getCreationalPolicy() {
    return creationalPolicy;
  }

  public List<EventFactory> getEventFactory() {
    return eventFactory;
  }

  public static class EventFactory {

    private String id;

    private EventMessageDefinitionData event;

    private List<EventFactoryModification> modification;

    protected EventFactory() {
    }

    public EventFactory(
        String id,
        EventMessageDefinitionData event,
        List<EventFactoryModification> modification
    ) {
      this.id = id;
      this.event = event;
      this.modification = modification;
    }

    public EventFactory(
        EventMessageDefinitionData event,
        List<EventFactoryModification> modification
    ) {
      this(
          UUID.randomUUID().toString(),
          event,
          modification
      );
    }

    public void addModification(EventFactoryModification modification) {
      this.getModification().add(modification);
    }

    public void addModifications(List<EventFactoryModification> modifications) {
      this.getModification().addAll(modification);
    }

    public EventMessageDefinitionData getEvent() {
      return event;
    }

    public List<EventFactoryModification> getModification() {
      return modification;
    }

    public String getId() {
      return id;
    }

    public static class EventFactoryModification {

      public static final String ADD = "add";
      public static final String UPSERT = "upsert";
      public static final String REMOVE = "remove";
      public static final String INSERT = "insert";
      public static final String MOVE = "move";
      private String kind;
      private String modificationPath;
      
      @Nullable
      private String startIdParameterName;

      @Nullable
      private String inputValueParameterName;

      @Nullable
      private String destinationIndexParameterName;

      private EventFactoryModification() {
      }

      private EventFactoryModification(
          String kind,
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName,
          @Nullable String destinationIndexParameterName
      ) {
        this.kind = kind;
        this.modificationPath = modificationPath;
        this.startIdParameterName = startIdParameterName;
        this.inputValueParameterName = inputValueParameterName;
        this.destinationIndexParameterName = destinationIndexParameterName;
      }

      public static EventFactoryModification add(
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName
      ) {
        return new EventFactoryModification(
            ADD,
            modificationPath,
            startIdParameterName,
            inputValueParameterName,
            null
        );
      }

      public static EventFactoryModification upsert(
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName
      ) {
        return new EventFactoryModification(
            UPSERT,
            modificationPath,
            startIdParameterName,
            inputValueParameterName,
            null
        );
      }

      public static EventFactoryModification remove(
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName
      ) {
        return new EventFactoryModification(
            REMOVE,
            modificationPath,
            startIdParameterName,
            inputValueParameterName,
            null
        );
      }

      public static EventFactoryModification insert(
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName,
          @Nullable String destinationIndexParameterName
      ) {
        return new EventFactoryModification(
            INSERT,
            modificationPath,
            startIdParameterName,
            inputValueParameterName,
            destinationIndexParameterName
        );
      }

      public static EventFactoryModification move(
          String modificationPath,
          @Nullable String startIdParameterName,
          @Nullable String inputValueParameterName,
          @Nullable String destinationIndexParameterName
      ) {
        return new EventFactoryModification(
            MOVE,
            modificationPath,
            startIdParameterName,
            inputValueParameterName,
            destinationIndexParameterName
        );
      }

      public String getKind() {
        return kind;
      }

      public String getModificationPath() {
        return modificationPath;
      }

      public String getStartIdParameterName() {
        return startIdParameterName;
      }

      @Nullable
      public String getInputValueParameterName() {
        return inputValueParameterName;
      }

      @Nullable
      public String getDestinationIndexParameterName() {
        return destinationIndexParameterName;
      }
    }
  }

  public static class CreationPolicy {

    public static final String ALWAYS = "always";
    public static final String NEVER = "never";
    public static final String IF_MISSING = "if_missing";

    private CreationPolicy() {
    }
  }
}
