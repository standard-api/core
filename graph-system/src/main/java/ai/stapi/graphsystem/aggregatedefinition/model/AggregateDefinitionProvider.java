package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import java.util.List;

public interface AggregateDefinitionProvider {

  List<AggregateDefinitionDTO> provideAll();

  AggregateDefinitionDTO provide(String aggregateType) throws CannotProvideAggregateDefinition;

  default AggregateDefinitionDTO getAggregateForOperation(
      String operationDefinitionId
  ) throws CannotProvideAggregateDefinition {
    var foundAggregates = this.provideAll().stream()
        .filter(aggregate -> aggregate.containsOperation(operationDefinitionId))
        .toList();

    if (foundAggregates.isEmpty()) {
      throw CannotProvideAggregateDefinition.becauseThereIsNoCommandHandlerForOperation(
          operationDefinitionId
      );
    }
    return foundAggregates.get(0);
  }

  default boolean containsAggregateForOperation(String operationDefinitionId) {
    return this.provideAll().stream().anyMatch(
        aggregate -> aggregate.containsOperation(operationDefinitionId)
    );
  }
}
