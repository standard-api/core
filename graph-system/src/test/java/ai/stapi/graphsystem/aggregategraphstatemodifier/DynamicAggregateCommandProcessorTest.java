package ai.stapi.graphsystem.aggregategraphstatemodifier;

import ai.stapi.graph.Graph;
import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.attribute.attributeValue.UriAttributeValue;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.aggregategraphstatemodifier.exceptions.CannotAddToAggregateState;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.systemfixtures.model.SystemModelDefinitionsLoader;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(SystemModelDefinitionsLoader.SCOPE)
class DynamicAggregateCommandProcessorTest extends SchemaIntegrationTestCase {

  @Autowired
  private DynamicAggregateCommandProcessor dynamicAggregateCommandProcessor;

  @Test
  public void itWillCreateMainNodeOnly() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of()
        ),
        new Graph()
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateWithPrimitiveParameters() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "implicitRules", "http://some.uri.com"
            )
        ),
        new Graph()
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateWithComplexParameter() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "meta", Map.of(
                    "source", "http://some.source.uri"
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateWithListParameter() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "parameter", List.of(
                    Map.of(
                        "name", "someParameterName1"
                    ),
                    Map.of(
                        "name", "someParameterName2"
                    ),
                    Map.of(
                        "name", "someParameterName3"
                    )
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateWithContentReferenceParameter() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "parameter", List.of(
                    Map.of(
                        "name", "basicParameterName"
                    ),
                    Map.of(
                        "name", "compositeParameter",
                        "part", List.of(
                            Map.of(
                                "name", "childParameter"
                            )
                        )
                    )
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateWithUnionParameter() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "parameter", List.of(
                    Map.of(
                        "name", "basicParameterName"
                    ),
                    Map.of(
                        "name", "parameterWithUnionName",
                        "value", Map.of(
                            "serializationType", "BoxedString",
                            "value", "SomeValue"
                        )
                    ),
                    Map.of(
                        "name", "anotherParameterWithUnionName",
                        "value", Map.of(
                            "serializationType", "Coding",
                            "code", "SomeCode"
                        )
                    )
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillCreateTargetNodeWhenAddingItemAndTargetNodeDoesntExist() {
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "AddParameterOnParameters",
            Map.of(
                "parameter", List.of(
                    Map.of("name", "someNewParameter")
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillAddItemOnAlreadyExistingGraphAndEventWillOnlyContainNewData() {
    var someId = new UniqueIdentifier("SomeId");
    var mainNode = new Node(someId, "Parameters");
    var alreadyExistingParameterNode = new Node(
        "ParametersParameter",
        new LeafAttribute<>("name", new StringAttributeValue("alreadyExistingParameter"))
    );
    var edge = new Edge(mainNode, "parameter", alreadyExistingParameterNode);
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "AddParameterOnParameters",
            Map.of(
                "parameter", List.of(
                    Map.of("name", "someNewParameter")
                )
            )
        ),
        new Graph(mainNode, alreadyExistingParameterNode, edge),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillThrowExceptionWhenStartIdParameterIsNotPresent() {
    var someId = new UniqueIdentifier("SomeId");
    Executable throwable = () -> this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "AddPartOnParametersParameter",
            Map.of(
                "part", List.of(
                    Map.of("name", "someNewDeeperParameter")
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, throwable);
  }

  @Test
  public void itWillThrowExceptionWhenStartIdIsNotOfCorrectFormat() {
    var someId = new UniqueIdentifier("SomeId");
    Executable throwable = () -> this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "AddPartOnParametersParameter",
            Map.of(
                "parameterId", "BadStartId",
                "part", List.of(
                    Map.of("name", "someNewDeeperParameter")
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, throwable);
  }

  @Test
  public void itWillNotAddItemOnDeeperNodeWhenStartNodeDoesntExist() {
    var someId = new UniqueIdentifier("SomeId");
    Executable throwable = () -> this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "AddPartOnParametersParameter",
            Map.of(
                "parameterId", "ParametersParameter/NonExistingParameterId",
                "part", List.of(
                    Map.of("name", "someNewDeeperParameter")
                )
            )
        ),
        new Graph(),
        MissingFieldResolvingStrategy.LENIENT
    );
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, throwable);
  }

  @Test
  public void itWillAddItemOnDeeperNodeAndOnlyIncludeNewDataInEvent() {
    var someId = new UniqueIdentifier("SomeId");
    var someParameterId = "SomeParameterId";
    var mainNode = new Node(someId, "Parameters");
    var startNode = new Node(
        new UniqueIdentifier(someParameterId),
        "ParametersParameter",
        new LeafAttribute<>("name", new StringAttributeValue("alreadyExistingParameter"))
    );
    var edge = new Edge(mainNode, "parameter", startNode);
    var actualEvent = this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "AddPartOnParametersParameter",
            Map.of(
                "parameterId", String.format("ParametersParameter/%s", someParameterId),
                "part", List.of(
                    Map.of("name", "someNewDeeperParameter")
                )
            )
        ),
        new Graph(mainNode, startNode, edge),
        MissingFieldResolvingStrategy.LENIENT
    );
    Assertions.assertEquals(1, actualEvent.size());
    this.thenGraphApproved(actualEvent.get(0).getSynchronizedGraph());
  }

  @Test
  public void itWillThrowExceptionWhenAddingParameterWhichAlreadyExists() {
    var someId = new UniqueIdentifier("SomeId");
    var mainNode = new Node(
        someId,
        "Parameters",
        new LeafAttribute<>("language", new StringAttributeValue("cz"))
    );
    Executable throwable = () -> this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            someId,
            "CreateParameters",
            Map.of(
                "language", "eng"
            )
        ),
        new Graph(mainNode)
    );
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, throwable);
  }

  @Test
  public void itWillThrowExceptionWhenAddingComplexTypeWhichAlreadyExists() {
    var someId = new UniqueIdentifier("SomeId");
    var mainNode = new Node(someId, "Parameters");
    var alreadyExistingComplex = new Node(
        "Meta",
        new LeafAttribute<>("source", new UriAttributeValue("http://some.already.existing.url"))
    );
    var edge = new Edge(mainNode, "meta", alreadyExistingComplex);
    Executable throwable = () -> this.dynamicAggregateCommandProcessor.processCommand(
        new DynamicCommand(
            new UniqueIdentifier("SomeId"),
            "CreateParameters",
            Map.of(
                "language", "cz",
                "meta", Map.of(
                    "source", "http://some.source.uri"
                )
            )
        ),
        new Graph(mainNode, alreadyExistingComplex, edge)
    );
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, throwable);
  }
}