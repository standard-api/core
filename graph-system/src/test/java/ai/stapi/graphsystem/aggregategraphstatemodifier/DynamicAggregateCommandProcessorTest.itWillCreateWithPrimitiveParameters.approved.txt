nodes:
	node_type: Parameters
	node_hash: D1B15
	node_edges:
	attributes:
		implicitRules -> http://some.uri.com
		language -> cz

edges:
